pkg/kubelet/kubelet.go

```diff
@@ -3076,3 +3061,79 @@ func (kl *Kubelet) warnCgroupV1Usage() {
 	}
 	metrics.CgroupVersion.Set(float64(cgroupVersion))
 }
+
+// staticPodRegistration ensures that all static pods are registered to the apiserver
+// before node became ready.
+func (kl *Kubelet) initialStaticPodsRegistration() error {
+	kl.initialStaticPodsRegisteredLock.Lock()
+	defer kl.initialStaticPodsRegisteredLock.Unlock()
+	// kubelet running in standalone mode does not register static pods to the apiserver.
+	if kl.kubeClient == nil {
+		return nil
+	}
+	// Check if initial static pod registration has already been completed.
+	if kl.initialStaticPodsRegistered {
+		return nil
+	}
+	staticPodToMirrorPodMap := kl.podManager.GetStaticPodToMirrorPodMap()
+	for _, mirrorPod := range staticPodToMirrorPodMap {
+		if mirrorPod == nil {
+			return fmt.Errorf("not all static pods are registered")
+		}
+	}
+
+	kl.initialStaticPodsRegistered = true
```

Comment:
we may want to add a metric to track the time until all the static pods are registered, it is simply to add just check https://github.com/kubernetes/kubernetes/pull/121041 for reference

Reply:
absolutely